File: Classes/CollapsibleTable.m
/*
	MyFlightbook for iOS - provides native access to MyFlightbook
	pilot's logbook
 Copyright (C) 2013-2023 MyFlightbook, LLC
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

//
//  CollapsibleTable.m
//  MFBSample
//
//  Created by Eric Berman on 3/19/13.
//
//

#import "CollapsibleTable.h"
#import "ExpandHeaderCell.h"
#import "PropertyCell.h"
#import "NavigableCell.h"
#import <UniformTypeIdentifiers/UniformTypeIdentifiers.h>

@import Photos;

@interface CollapsibleTable ()
@end

@implementation CollapsibleTable

BOOL fSelectActiveSelOnScrollCompletion = NO;
BOOL fSelectFirst = NO;

@synthesize expandedSections, ipActive, fIsValid, defSectionFooterHeight, defSectionHeaderHeight;

#pragma mark - View Creation/Destruction
- (void) setUpCollapsibleTable
{
    self.expandedSections = [[NSMutableIndexSet alloc] init];
    self.fIsValid = NO;
    self.defSectionHeaderHeight = self.defSectionFooterHeight = 1;
}

- (instancetype)initWithStyle:(UITableViewStyle)style
{
    self = [super initWithStyle:UITableViewStyleGrouped];
    if (self) {
        [self setUpCollapsibleTable];
    }
    return self;
}

- (instancetype) init
{
    self = [super init];
    [self setUpCollapsibleTable];
    return self;
}

- (instancetype) initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    [self setUpCollapsibleTable];
    return self;
}

- (instancetype) initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    [self setUpCollapsibleTable];
    return self;
}

- (void) viewDidLoad
{
    [super viewDidLoad];

    // Fix an iOS 11 issue with going under the UITabBar?
    if ([self respondsToSelector:@selector(edgesForExtendedLayout)])
        self.edgesForExtendedLayout = UIRectEdgeNone;
}

- (void) viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
}

- (void) viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
}

#pragma mark - Misc ViewController
- (BOOL) shouldAutorotate
{
    return YES;
}

- (UIInterfaceOrientationMask) supportedInterfaceOrientations
{
    return UIInterfaceOrientationMaskAll;
}

- (void) reload {
    [self.tableView reloadData];
}

#pragma mark - Expand/collapse support
- (void) collapseSection:(NSInteger) section forTable:(UITableView *) tv
{
    if (![self isExpanded:section])
        return;

    NSInteger oldRowCount = [self tableView:tv numberOfRowsInSection:section];
    [self.expandedSections removeIndex:section];
    ExpandHeaderCell * cell = (ExpandHeaderCell *) [tv cellForRowAtIndexPath:[NSIndexPath indexPathForRow:0 inSection:section]];
    [cell setExpanded:NO];
    
    NSMutableArray * rg = [[NSMutableArray alloc] initWithCapacity:oldRowCount];
    for (NSInteger i = oldRowCount - 1; i >= 1; i--)
        [rg addObject:[NSIndexPath indexPathForRow:i inSection:section]];
    [tv deleteRowsAtIndexPaths:rg withRowAnimation:UITableViewRowAnimationTop];
}

- (void) expandSection:(NSInteger) section forTable:(UITableView *) tv
{
    if ([self isExpanded:section])
        return;

    [self.expandedSections addIndex:section];
    ExpandHeaderCell * cell = (ExpandHeaderCell *) [tv cellForRowAtIndexPath:[NSIndexPath indexPathForRow:0 inSection:section]];
    [cell setExpanded:YES];
    
    NSInteger newRowCount = [self tableView:tv numberOfRowsInSection:section];
    NSMutableArray * rg = [[NSMutableArray alloc] initWithCapacity:newRowCount];
    for (int i = 1; i < newRowCount; i++)
        [rg addObject:[NSIndexPath indexPathForRow:i inSection:section]];
    [tv insertRowsAtIndexPaths:rg withRowAnimation:UITableViewRowAnimationTop];
}

- (void) toggleSection:(NSInteger) section forTable:(UITableView *) tableView
{
    if ([self.expandedSections containsIndex:section])
        [self collapseSection:section forTable:tableView];
    else
        [self expandSection:section forTable:tableView];
}

- (void) collapseSection:(NSInteger) section
{
    [self collapseSection:section forTable:self.tableView];
}

- (void) expandSection:(NSInteger) section
{
    [self expandSection:section forTable:self.tableView];
}

- (void) toggleSection:(NSInteger) section
{
    [self toggleSection:section forTable:self.tableView];
}

- (BOOL) isExpanded:(NSInteger) section
{
    return [self.expandedSections containsIndex:section];
}

- (void) collapseAll
{
    [self.expandedSections removeAllIndexes];
}

- (void) expandAll:(UITableView *) tableView
{
    NSInteger cSections = [tableView numberOfSections];
    for (NSInteger i = 0; i < cSections; i++)
        [self.expandedSections addIndex:i];
}

#pragma mark - Basic Accessorybar Delegate Support
- (UITableViewCell *) owningCellGeneric:(UIView *)vw
{
    UITableViewCell * tc = nil;
    while (vw != nil)
    {
        vw = vw.superview;
        if ([vw isKindOfClass:[UITableViewCell class]])
            return (UITableViewCell *) vw;
    }
    return tc;
}

- (EditCell *) owningCell:(UIView *) vw
{
    EditCell * ec = nil;
    
    while (vw != nil)
    {
        vw = vw.superview;
        if ([vw isKindOfClass:[EditCell class]])
            return (EditCell *) vw;
    }
    
    return ec;
}

- (BOOL) isNavigableRow:(NSIndexPath *) ip
{
    // this should be overriden.
    return false;
}

- (NSIndexPath *) nextNavCell:(NSIndexPath *) ip
{
    NSIndexPath * ipNext = [self nextCell:ip];
    while (ipNext.row != ip.row || ipNext.section != ip.section)
    {
        if ([self isNavigableRow:ipNext])
            return ipNext;
        ip = ipNext;
        ipNext = [self nextCell:ip];
    }
    return nil;
}

- (NSIndexPath *) prevNavCell:(NSIndexPath *) ip
{
    NSIndexPath * ipPrev = [self prevCell:ip];
    while (ipPrev.row != ip.row || ipPrev.section != ip.section)
    {
        if ([self isNavigableRow:ipPrev])
            return ipPrev;
        ip = ipPrev;
        ipPrev = [self prevCell:ip];
    }
    return nil;
}

- (BOOL) canNext
{
    return [self nextNavCell:self.ipActive] != nil;
}

- (BOOL) canPrev
{
    return [self prevNavCell:self.ipActive] != nil;
}

- (void) enableNextPrev:(AccessoryBar *) vwAccessory
{
    vwAccessory.btnNext.enabled = [self canNext];
    vwAccessory.btnPrev.enabled = [self canPrev];
}

- (void) scrollViewDidEndScrollingAnimation:(UIScrollView *)scrollView
{
    if (fSelectActiveSelOnScrollCompletion)
    {
        UITableViewCell * cell = [self.tableView cellForRowAtIndexPath:self.ipActive];
        if (cell != nil)
        {
            if ([cell isKindOfClass:[NavigableCell class]])
            {
                NavigableCell * nc = (NavigableCell *) cell;
                UIResponder * resp = (fSelectFirst) ? nc.firstResponderControl : nc.lastResponderControl;
                if (resp.canBecomeFirstResponder)
                    [resp becomeFirstResponder];
            }
        }
    }
    fSelectActiveSelOnScrollCompletion = NO;
}

- (void) navigateToActiveCell
{
    fSelectActiveSelOnScrollCompletion = YES;
    [self.tableView scrollToRowAtIndexPath:self.ipActive atScrollPosition:UITableViewScrollPositionMiddle animated:YES];
    // Scrolling MIGHT NOT happen if the cell is visible.
    // So test to see if the tableview already has it.  If so, we can call scrollViewDidEndScrollingAnimation ourselves
    if ([self.tableView cellForRowAtIndexPath:self.ipActive] != nil)
        [self scrollViewDidEndScrollingAnimation:self.tableView];
}

- (void) nextClicked
{
    NSIndexPath * ipNext = [self nextNavCell:self.ipActive];
    if (ipNext != nil)
    {
        NSLog(@"Navigating from %ld,%ld to %ld,%ld", (long) self.ipActive.section, (long) self.ipActive.row, (long) ipNext.section, (long)ipNext.row);
        self.ipActive = ipNext;
        fSelectFirst = YES;
        [self navigateToActiveCell];
    }
}

- (void) prevClicked
{
    NSIndexPath * ipPrev = [self prevNavCell:self.ipActive];
    if (ipPrev != nil)
    {
        NSLog(@"Navigating from %ld,%ld to %ld,%ld", (long) self.ipActive.section, (long) self.ipActive.row, (long) ipPrev.section, (long) ipPrev.row);
        self.ipActive = ipPrev;
        fSelectFirst = NO;
        [self navigateToActiveCell];
    }
}

- (void) doneClicked
{
    [self.tableView endEditing:YES];
}

- (void) deleteClicked
{
    EditCell * ec = (EditCell *) [self.tableView cellForRowAtIndexPath:self.ipActive];
    ec.txt.text = ec.txtML.text = @"";
}

// Issue #284 - trap any physical keyboard tab events.
- (void) pressesBegan:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event {
    for (UIPress * press in presses) {
        if (press.key != nil && press.key.keyCode == UIKeyboardHIDUsageKeyboardTab) {
            if (press.key.modifierFlags == UIKeyModifierShift) {
                if (self.canPrev)
                    [self prevClicked];
            }
            else {
                if (self.canNext)
                    [self nextClicked];
            }
            return;
        }
    }
    [super pressesBegan:presses withEvent:event];
}

#pragma mark - Camera support
- (BOOL) canUseCamera
{
    return [UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera];
}

- (BOOL) canUsePhotoLibrary
{
    return [UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypePhotoLibrary];
}

- (void) addImages:(BOOL)usingCamera fromButton:(id)btn
{
	UIImagePickerController * imgView = [[UIImagePickerController alloc] init];
    
	imgView.delegate = self;
    
    BOOL fIsCameraAvailable = self.canUseCamera;
    BOOL fIsGalleryAvailable = self.canUsePhotoLibrary;
    if ((usingCamera && !fIsCameraAvailable) || (!usingCamera && !fIsGalleryAvailable))
        return;
    
    imgView.sourceType = (usingCamera && fIsCameraAvailable) ? UIImagePickerControllerSourceTypeCamera : UIImagePickerControllerSourceTypePhotoLibrary;
    imgView.mediaTypes =  [UIImagePickerController availableMediaTypesForSourceType:imgView.sourceType];
    
    imgView.modalPresentationStyle = UIModalPresentationPopover;
    UIPopoverPresentationController * ppc = imgView.popoverPresentationController;
    ppc.barButtonItem = btn;
    ppc.permittedArrowDirections = UIPopoverArrowDirectionAny;
    ppc.delegate = self;
    [self presentViewController:imgView animated:YES completion:^{}];
}

- (IBAction) pickImages:(id) sender
{
    // Request permission so that we can get geotag
    // Technically not required, so we'll call addImages regardless.
    PHAuthorizationStatus status = [PHPhotoLibrary authorizationStatus];
    switch (status) {
        case PHAuthorizationStatusNotDetermined: {
            [PHPhotoLibrary requestAuthorization:^(PHAuthorizationStatus status) {
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self addImages:NO fromButton:sender];
                });
            }];
        }
            break;
        case PHAuthorizationStatusDenied:
        case PHAuthorizationStatusRestricted:
        case PHAuthorizationStatusLimited:
        case PHAuthorizationStatusAuthorized:
            [self addImages:NO fromButton:sender];
            break;
    }
}

- (IBAction) takePicture:(id) sender
{
	[self addImages:YES fromButton:sender];
}

#pragma mark - UIImagePickerControllerDelegate
- (void) stopPickingPictures
{
	[self dismissViewControllerAnimated:YES completion:^{}];
	[self.tableView reloadData];
}

- (void) addImage:(CommentedImage *)ci
{
    // Subclass this to add an image!!
}

- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary *)info
{
    NSString * szType = info[UIImagePickerControllerMediaType];

    BOOL fImage = [szType compare:UTTypeImage.identifier] == NSOrderedSame;
    BOOL fVideo = [szType compare:UTTypeMovie.identifier] == NSOrderedSame;
    BOOL fCamera = (picker.sourceType == UIImagePickerControllerSourceTypeCamera);
    if (fImage || fVideo)
    {
        CLLocation * loc = nil;
        PHAsset * thisPhoto = nil;
        if (!fCamera) {
            thisPhoto = info[UIImagePickerControllerPHAsset];
            if (thisPhoto != nil)
                loc = thisPhoto.location;
        }
        
        CommentedImage * ci = [[CommentedImage alloc] init];
        if (fImage)
        {
            if (!fCamera && loc != nil)
                ci.imgInfo.Location = [[MFBWebServiceSvc_LatLong alloc] initWithCoord:loc.coordinate];

            UIImage * img = info[UIImagePickerControllerOriginalImage];
            [ci SetImage:img fromCamera:fCamera withMetaData:info];
        }
        else if (fVideo)
        {
            [ci SetVideo:info[UIImagePickerControllerMediaURL] fromCamera:fCamera];
        }

        [self addImage:ci];
        [self stopPickingPictures];
    }
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker
{
	[self stopPickingPictures];
}

#pragma mark - UINavigationControllerDelegate
- (void)navigationController:(UINavigationController *)navigationController didShowViewController:(UIViewController *)viewController animated:(BOOL)animated
{
}

- (void)navigationController:(UINavigationController *)navigationController willShowViewController:(UIViewController *)viewController animated:(BOOL)animated
{
	if (viewController == self)
		[self.tableView reloadData];
}

#pragma mark - Background Image
- (void) setBackground:(NSString *) szImageName
{
    self.tableView.backgroundView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:szImageName]];
    self.tableView.backgroundView.contentMode = UIViewContentModeTopLeft;
}

#pragma mark - iOS7 hacks
// Reduce the space between sections.
- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section
{
    CGFloat height = self.defSectionFooterHeight;    // default
    NSString * sz = [self tableView:tableView titleForFooterInSection:section];
    if (tableView == self.tableView && [sz length] > 0)
    {
        CGFloat h = [sz boundingRectWithSize:CGSizeMake(self.tableView.frame.size.width - 20, 10000)
                                                        options:NSStringDrawingUsesLineFragmentOrigin
                                                     attributes:@{NSFontAttributeName:[UIFont systemFontOfSize:22.0]}
                                                        context:nil].size.height;
        return ceil(h);
    }
    return height;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section
{
    CGFloat height = self.defSectionHeaderHeight;
    NSString * sz = [self tableView:tableView titleForHeaderInSection:section];
    if (tableView == self.tableView && [sz length] > 0)
    {
        CGFloat h = [sz boundingRectWithSize:CGSizeMake(self.tableView.frame.size.width - 20, 10000)
                                     options:NSStringDrawingUsesLineFragmentOrigin
                                  attributes:@{NSFontAttributeName:[UIFont systemFontOfSize:22.0]}
                                     context:nil].size.height;
        return ceil(h);
    }
    
    return height;
}

#pragma mark - invalidate (for sign-out)
- (void) invalidateViewController
{
    // Should be subclassed
}
@end

File: Classes/CollapsibleTable.swift
/*
    MyFlightbook for iOS - provides native access to MyFlightbook
    pilot's logbook
 Copyright (C) 2013-2023 MyFlightbook, LLC
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

//
//  CollapsibleTable.swift
//  MyFlightbook
//
//  Created by Eric Berman on 3/26/23.
//

import Foundation
import Photos

/// Swift version of the CollapsibleTable class
/// MUST coexist with CollapsibleTable during swift transition because it relies on Swift protocols/classes
/// Since objc-classes can't inherit from a swift class, we can't migrate child classes unless/until
/// this is migrated without including collapsibletable.h in the bridging header, but if we do that, we get
/// circular references.
/// So this will duplicate the objc class, and swift classes can inherit from this.
/// We will do the same for pullrefreshtableviewcontroller

public class CollapsibleTableSw : UITableViewController, UIImagePickerControllerDelegate, AccessoryBarDelegate, Invalidatable, UINavigationControllerDelegate, UIPopoverPresentationControllerDelegate {
    public var expandedSections : Set<Int> = []
    public var ipActive : IndexPath? = nil
    public var fIsValid = false
    public var defSectionHeaderHeight : CGFloat = 1
    public var defSectionFooterHeight : CGFloat = 1
    
    private var fSelectActiveSelOnScrollCompletion = false
    private var fSelectFirst = false
    
    // MARK: Lifecycle
    // No need to do viewdidload/etc since all initialization is done above
    
    // MARK: Misc ViewController
    public override var shouldAutorotate: Bool {
        get {
            return true
        }
    }
    
    public override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        return .all
    }
    
    public func reload() {
        tableView.reloadData()
    }
    
    // MARK: Expand/collapse support
    public func collapseSection(_ section : Int, table tv : UITableView) {
        if !isExpanded(section) {
            return
        }
        
        let oldRowCount = tableView(tv, numberOfRowsInSection: section)
        expandedSections.remove(section)
        
        // cellforrow can return nil if the cell isn't visible yet, which might be true in viewDidLoad
        let cell = tv.cellForRow(at: IndexPath(row: 0, section: section)) as? ExpandHeaderCell
        cell?.setExpanded(false)
        
        var rg : [IndexPath] = []
        for i in 1..<oldRowCount {
            rg.append(IndexPath(row: i, section: section))
        }
        tv.deleteRows(at: rg.reversed(), with: .top)
    }
    
    public func expandSection(_ section : Int, table tv : UITableView) {
        if isExpanded(section) {
            return
        }

        expandedSections.insert(section)

        // cellforrow can return nil if the cell isn't visible yet, which might be true in viewDidLoad
        let cell = tv.cellForRow(at: IndexPath(row: 0, section: section)) as? ExpandHeaderCell
        cell?.setExpanded(true)
        
        let newRowCount = tableView(tv, numberOfRowsInSection: section)
        var rg : [IndexPath] = []
        for i in 1..<newRowCount {
            rg.append(IndexPath(row: i, section: section))
        }
        tv.insertRows(at: rg, with: .top)
    }
    
    @objc public func toggleSection(_ section : Int, forTable tableView : UITableView) {
        if expandedSections.contains(section) {
            collapseSection(section, table: tableView)
        } else {
            expandSection(section, table: tableView)
        }
    }
    
    public func collapseSection(_ section : Int) {
        collapseSection(section, table: tableView)
    }
    
    public func expandSection(_ section : Int) {
        expandSection(section, table: tableView)
    }
    
    public func toggleSection(_ section : Int) {
        toggleSection(section, forTable: tableView)
    }
    
    public func isExpanded(_ section : Int) -> Bool {
        return expandedSections.contains(section)
    }
    
    public func collapseAll() {
        expandedSections.removeAll()
    }
    
    public func expandAll(_ tableView : UITableView) {
        let cSections = tableView.numberOfSections
        for i in 0..<cSections {
            expandedSections.insert(i)
        }
    }
    
    // MARK: Basic Accessorybar Delegate Support
    public func owningCellGeneric(_ vw : UIView) -> UITableViewCell? {
        var v : UIView? = vw
        while v != nil {
            v = v!.superview
            if let cell = v as? UITableViewCell {
                return cell
            }
        }
        return nil
    }
    
    public func owningCell(_ vw : UIView) -> EditCell? {
        var v : UIView? = vw
        while v != nil {
            v = v!.superview
            if let cell = v as? EditCell {
                return cell
            }
        }
        return nil
    }
    
    public func isNavigableRow(_ ip : IndexPath) -> Bool {
        // this should be overridden
        return false
    }
    
    func nextNavCell(_ ipIn : IndexPath?) -> IndexPath? {
            if ipIn == nil {
            return nil
        }
        var ip = ipIn!
        
        var ipNext = nextCell(ipCurrent: ip)
        while ipNext.row != ip.row || ipNext.section != ip.section {
            if isNavigableRow(ipNext) {
                return ipNext
            }
            ip = ipNext
            ipNext = nextCell(ipCurrent: ip)
        }
        return nil
    }
    
    func prevNavCell(_ ipIn : IndexPath?) -> IndexPath? {
        if ipIn == nil {
            return nil
        }
        var ip = ipIn!
        
        var ipPrev = prevCell(ipCurrent: ip)
        while ipPrev.row != ip.row || ipPrev.section != ip.section {
            if isNavigableRow(ipPrev) {
                return ipPrev
            }
            ip = ipPrev
            ipPrev = prevCell(ipCurrent: ip)
        }
        return nil
    }
    
    func canNext() -> Bool {
        return nextNavCell(ipActive) != nil
    }
    
    func canPrev() -> Bool {
        return prevNavCell(ipActive) != nil
    }
    
    public func enableNextPrev(_ vwAccesory : AccessoryBar) {
        vwAccesory.btnNext.isEnabled = canNext()
        vwAccesory.btnPrev.isEnabled = canPrev()
    }
    
    public override func scrollViewDidEndScrollingAnimation(_ scrollView: UIScrollView) {
        if fSelectActiveSelOnScrollCompletion && ipActive != nil {
            if let cell = tableView.cellForRow(at: ipActive!) {
                if let nc = cell as? NavigableCell {
                    let resp = fSelectFirst ? nc.firstResponderControl : nc.lastResponderControl
                    if resp?.canBecomeFirstResponder ?? false {
                        resp?.becomeFirstResponder()
                    }
                }
            }
        }
        fSelectActiveSelOnScrollCompletion = false
    }
    
    public func navigateToActiveCell() {
        if ipActive == nil {
            return
        }
        fSelectActiveSelOnScrollCompletion = true
        tableView.scrollToRow(at: ipActive!, at: .middle, animated: true)
        // Scrolling MIGHT NOT happen if the cell is visible.
        // So test to see if the tableview already has it.  If so, we can call scrollViewDidEndScrollingAnimation ourselves
        if tableView.cellForRow(at: ipActive!) != nil {
            scrollViewDidEndScrollingAnimation(tableView)
        }
    }
    
    public func nextClicked() {
        if let ipNext = nextNavCell(ipActive) {
            NSLog("Navigating from %d,%d to %d,%d", ipActive!.section,  ipActive!.row,  ipNext.section, ipNext.row)
            ipActive = ipNext
            fSelectFirst = true
            navigateToActiveCell()
        }
    }
    
    public func prevClicked() {
        if let ipPrev = prevNavCell(ipActive) {
            NSLog("Navigating from %d,%d to %d,%d", ipActive!.section,  ipActive!.row,  ipPrev.section, ipPrev.row)
            ipActive = ipPrev
            fSelectFirst = false
            navigateToActiveCell()
        }
    }
    
    public func doneClicked() {
        tableView.endEditing(true)
    }
    
    public func deleteClicked() {
        if ipActive != nil, let ec = tableView.cellForRow(at: ipActive!) as? EditCell {
            ec.txt?.text = ""
            ec.txtML?.text = ""
        }
    }
    
    // Issue #284 - trap any physical keyboard tab events.
    public override func pressesBegan(_ presses: Set<UIPress>, with event: UIPressesEvent?) {
        for press in presses {
            if press.key?.keyCode == .keyboardTab {
                if (press.key?.modifierFlags == .shift) {
                    if canPrev() {
                        prevClicked()
                    }
                }
                else {
                    if canNext() {
                        nextClicked()
                    }
                }
                return
            }
        }
        super.pressesBegan(presses, with: event)
    }
    
    // MARK: Camera support
    public func canUseCamera() -> Bool {
        return UIImagePickerController.isSourceTypeAvailable(.camera)
    }
    
    // TODO: Migrate to PHPicker from UIImagePicker?
    public func canUsePhotoLibrary() -> Bool {
        return UIImagePickerController.isSourceTypeAvailable(.photoLibrary)
    }
    
    func addImages(_ usingCamera : Bool, fromButton btn: UIBarButtonItem?) {
        let imgView = UIImagePickerController()
        imgView.delegate = self
        
        let fIsCameraAvailable = canUseCamera()
        let fIsGalleryAvailable = canUsePhotoLibrary()
        if usingCamera && !fIsCameraAvailable || !usingCamera && !fIsGalleryAvailable {
            return
        }
        
        imgView.sourceType = usingCamera && fIsCameraAvailable ? .camera : .photoLibrary
        imgView.mediaTypes = UIImagePickerController.availableMediaTypes(for: imgView.sourceType) ?? []
        
        imgView.modalPresentationStyle = .popover
        let ppc = imgView.popoverPresentationController
        ppc?.barButtonItem = btn
        ppc?.delegate = self
        present(imgView, animated: true)
    }
    
    @IBAction public func pickImages(_ sender : Any) {
        // Request permission so that we can get geotag
        // Technically not required, so we'll call addImages regardless.
        let status = PHPhotoLibrary.authorizationStatus()
        switch (status) {
        case .notDetermined:
            PHPhotoLibrary.requestAuthorization(for: .readWrite) { status in
                if status == .authorized {
                    DispatchQueue.main.async {
                        self.addImages(false, fromButton: sender as? UIBarButtonItem)
                    }
                }
            }
            case .denied, .restricted, .limited:
                break
        case .authorized:
            addImages(false, fromButton: sender as? UIBarButtonItem)
        default:
            break
        }
    }
    
    @IBAction public func takePictures(_ sender : Any) {
        addImages(true, fromButton: sender as? UIBarButtonItem)
    }
    
    // MARK: UIImagePickerControllerDelegate
    func stopPickingPictures() {
        dismiss(animated: true)
        tableView.reloadData()
    }
    
    public func addImage(_ ci : CommentedImage) {
        fatalError("addImage called directly on CollapsibleTable - Should be subclassed")
    }
    
    public func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        let szType = info[.mediaType] as? String
                
        let fImage = szType == UTType.image.identifier
        let fVideo = szType == UTType.movie.identifier
        let fCamera = picker.sourceType == .camera
        if (fImage || fVideo) {
            var loc : CLLocation? = nil
            if !fCamera {
                // Get the location of the library image
                if let thisPhoto = info[.phAsset] as? PHAsset {
                    loc = thisPhoto.location
                }
            }
            
            let ci = CommentedImage()
            if (fImage) {
                if !fCamera && loc != nil {
                    ci.imgInfo?.location = MFBWebServiceSvc_LatLong(coord: loc!.coordinate)
                }
                
                if let img = info[.originalImage] as? UIImage {
                    ci.SetImage(img, fromCamera: fCamera, withMetaData: info)
                }
            } else if (fVideo) {
                ci.SetVideo(info[.mediaURL] as! URL, fromCamera: fCamera)
            }
            
            addImage(ci)
            stopPickingPictures()
        }
    }
    
    public func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        stopPickingPictures()
    }
    
    // MARK: UINavigationControllerDelegate
    public func navigationController(_ navigationController: UINavigationController, didShow viewController: UIViewController, animated: Bool) {
        
    }
    
    public func navigationController(_ navigationController: UINavigationController, willShow viewController: UIViewController, animated: Bool) {
        if viewController == self {
            tableView.reloadData()
        }
    }
    
    // MARK: Background image
    public func setBackground(_ szImageName : String) {
        tableView.backgroundView = UIImageView(image: UIImage(named: szImageName))
        tableView.backgroundView?.contentMode = .topLeft
    }
    
    // MARK: iOS7 hacks
    // Reduce the space between sections.
    public override func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        let sz = self.tableView(tableView, titleForFooterInSection:section) ?? ""
        
        if tableView == tableView && !sz.isEmpty {
            let h = (sz as NSString).boundingRect(with: CGSizeMake(self.tableView.frame.size.width - 20, 10000),
                                                  options:.usesLineFragmentOrigin,
                                                  attributes: [.font : UIFont.systemFont(ofSize: 22.0)],
                                                  context: nil).size.height
            return ceil(h)
        }
        return defSectionFooterHeight
    }
    
    public override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        let sz = self.tableView(tableView, titleForHeaderInSection:section) ?? ""
        
        if tableView == tableView && !sz.isEmpty {
            let h = (sz as NSString).boundingRect(with: CGSizeMake(self.tableView.frame.size.width - 20, 10000),
                                                  options:.usesLineFragmentOrigin,
                                                  attributes: [.font : UIFont.systemFont(ofSize: 22.0)],
                                                  context: nil).size.height
            return ceil(h)
        }
        return defSectionHeaderHeight
    }
    
    // MARK: Invalidate (for sign-out)
    public func invalidateViewController() {
        fatalError("invalidatedViewController called directly on CollapsibleTable - Should be subclassed")
    }
}
 
